### DISCUSSION TITLE
not_allowed plugin usage with `r.root`?

### DISCUSSION THREAD

**Discussion by jgarber623** [jeremyevans/roda#266] [General]
Hello! I'm relatively new to Roda, so apologies if this has come up in conversation before. I did a quick search of code and discussions and didn't find anything, so here goes…

Is the not_allowed plugin ([documentation](http://roda.jeremyevans.net/rdoc/classes/Roda/RodaPlugins/NotAllowed.html)) supposed to work with routes defined using `r.root`? I'm seeing some confusing behavior with even a modest app:

```rb
# config.ru

require 'roda'

class App < Roda
  plugin :not_allowed

  route do |r|
    r.root do
      'Rendered r.root!'
    end
  end
end

run App.freeze.app
```

The above should yield an app with a single route responding at `GET /`. Run that with `rackup` and:

```sh
# Response body of 'Rendered r.root!'
curl http://localhost:9292

# 404…? Expected a 405 here since HEAD is not defined in App
curl --head http://localhost:9292
```

If `App` is updated to use `r.get` instead of `r.root`:

```rb
r.get '' do
  'Rendered get!'
end
```

…then `curl --head http://localhost:9292` properly returns:

```
HTTP/1.1 405 Method Not Allowed
Allow: GET
Content-Type: text/html
Content-Length: 3
```

I get similar results for POST requests, etc.

So, question: **Is this behavior intentional?** Or, should `r.root` behave similarly to `r.get ''`?

**Comment by jeremyevans** [comment]
The not_allowed plugin previously didn't override `r.root` to return 405 responses for non-GET requests.  I added support for that in 71e293824aea1841046332dd461185efc2bf985f .  So with that change, you'll get the 405 response for HEAD request.

Note that if you want to have HEAD requests treated as GET requests without bodies, you can use the head plugin for that.  As documented in the head plugin, you want to load it after the not_allowed plugin.

**Reply by jgarber623** [reply]
Thanks so much, @jeremyevans! Appreciate the quick update and thoughtful response.

Good note on the use of the `head` plugin, too. I should've made it clearer in my original post that the issue existed across other HTTP verbs, too. I used `curl --head` in the examples for expediency (`curl -d '{}' http://localhost:9292` should've demonstrated the same behavior).

**Comment by jgarber623** [comment]
Noting that this was released with v3.56.0 (#271). 🎉
